crates/bevy_asset/src/loader_builders.rs

```diff
@@ -25,28 +25,163 @@ impl ReaderRef<'_> {
+/// as possible. "Normal loads" kicked from within a normal Bevy App will
+/// generally configure the context to kick off loads immediately.
+///
+/// If the current context is configured to not load dependencies automatically
+/// (ex: [`AssetProcessor`]), a load will not be kicked off automatically. It is
+/// then the calling context's responsibility to begin a load if necessary.
 ///
 /// # Lifetimes
+///
 /// - `ctx`: the lifetime of the associated [`AssetServer`](crate::AssetServer) reference
 /// - `builder`: the lifetime of the temporary builder structs
-pub struct NestedLoader<'ctx, 'builder> {
+///
+/// [`deferred`]: Self::deferred
+/// [`immediate`]: Self::immediate
+/// [`load`]: Self::load
+/// [`with_dynamic_type`]: Self::with_dynamic_type
+/// [`AssetServer::load`]: crate::AssetServer::load
+/// [`AssetProcessor`]: crate::processor::AssetProcessor
+/// [`Process`]: crate::processor::Process
+/// [`LoadTransformAndSave`]: crate::processor::LoadTransformAndSave
+pub struct NestedLoader<'ctx, 'builder, T, M> {
     load_context: &'builder mut LoadContext<'ctx>,
     meta_transform: Option<MetaTransform>,
-    asset_type_id: Option<TypeId>,
+    typing: T,
+    mode: M,
+}
+
+mod sealed {
+    pub trait Typing {}
+
+    pub trait Mode {}
+}
+
+/// [`NestedLoader`] will be provided the type of asset as a type parameter on
+/// [`load`].
+///
+/// [`load`]: NestedLoader::load
+pub struct StaticTyped {
+    _priv: (),
 }
 
-impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
-    pub(crate) fn new(
-        load_context: &'builder mut LoadContext<'ctx>,
-    ) -> NestedLoader<'ctx, 'builder> {
+impl sealed::Typing for StaticTyped {}
+
+/// [`NestedLoader`] has been configured with info on what type of asset to load
+/// at runtime.
+pub struct DynamicTyped {
+    asset_type_id: TypeId,
+}
+
+impl sealed::Typing for DynamicTyped {}
+
+/// [`NestedLoader`] does not know what type of asset it will be loading.
+pub struct UnknownTyped {
+    _priv: (),
+}
+
+impl sealed::Typing for UnknownTyped {}
+
+/// [`NestedLoader`] will create and return asset handles immediately, but only
+/// actually load the asset later.
+pub struct Deferred {
+    _priv: (),
+}
+
+impl sealed::Mode for Deferred {}
+
+/// [`NestedLoader`] will immediately load an asset when requested.
+pub struct Immediate<'builder, 'reader> {
+    reader: Option<&'builder mut (dyn Reader + 'reader)>,
+}
+
+impl sealed::Mode for Immediate<'_, '_> {}
+
+// common to all states
+
+impl<'ctx, 'builder> NestedLoader<'ctx, 'builder, StaticTyped, Deferred> {
+    pub(crate) fn new(load_context: &'builder mut LoadContext<'ctx>) -> Self {
         NestedLoader {
             load_context,
             meta_transform: None,
-            asset_type_id: None,
+            typing: StaticTyped { _priv: () },
+            mode: Deferred { _priv: () },
         }
     }
+}
 
+impl<'ctx, 'builder, T: sealed::Typing, M: sealed::Mode> NestedLoader<'ctx, 'builder, T, M> {
```

Comment:
One thing that's a little odd to me, I'm not sure if we care enough: with this impl, you can say:

```rust
load_context.loader().with_static_type().with_static_type().with_static_type()...
```

We could solve this with separate impl blocks, but perhaps that's too much boilerplate.

Feel free to ack!

Reply:
I think this small improvement would be nice. I bet you someone will notice it and report an issue anyway!

Reply:
To enable this for at least just the `T` parameter, I'd need:
```rs
impl<M> NestedLoader<StaticTyped, M> {
    fn with_dynamic_type() -> .. {}
    fn with_unknown_type() -> .. {}
}

impl<M> NestedLoader<DynamicTyped, M> {
    fn with_static_type() -> .. {}
    fn with_unknown_type() -> .. {}
}

impl<M> NestedLoader<UnknownType, M> {
    fn with_static_type() -> .. {}
    fn with_dynamic_type -> .. {}
}
```
I really don't like the duplication here, but what's a deal breaker for me with this approach IMO is I would have to duplicate docs for both of the `with_x_type` etc. Docs are so much harder to keep up to date than code, and I'd always prioritise having a single source of truth for docs over duplicating docs. Inevitably, if we do this, someone will change how these functions work and change the docs for one of these, then forget the other.

I could also add some sealed traits like `TypedNotStatic`, `TypedNotDynamic`, `TypedNotUnknown`, etc. and impl for those, but IMO this is over-engineering a solution to something that isn't really a problem. I think if someone writes
```rs
my_loader
    .with_static_type()
    .with_static_type()
```
it's not really harmful/a footgun of any kind, and it's easily spotted in code review.

The same problem exists for `M` as well, but to a lesser extent.